tests/extension/oneapi_weak_object/weak_object_common.h

```diff
@@ -0,0 +1,393 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Copyright (c) 2023 The Khronos Group Inc.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+*******************************************************************************/
+
+#ifndef __SYCLCTS_TESTS_WEAK_OBJECT_COMMON_H
+#define __SYCLCTS_TESTS_WEAK_OBJECT_COMMON_H
+
+#include "../../../util/sycl_exceptions.h"
+#include "../../common/common.h"
+#include "../../common/type_coverage.h"
+
+#ifdef SYCL_EXT_ONEAPI_WEAK_OBJECT
+
+namespace weak_object_common {
+
+template <typename SYCLObjT>
+SYCLObjT get_sycl_object() {
+  static sycl::buffer<int> buf{{1}};
```

Comment:
It should be enough to just have it select different buffers in the case where you call this function twice. The problem only happens because it depends on the same memory.

Reply:
I have just taken your changes from fix-patches and all tests was passed. So, is it defined behavior with `host_accesor` or just a bug?

Reply:
Reviewing the specification, it states that

> Host Accessors: The constructor for a host accessor waits for all kernels that modify the same buffer (or image) in any queues to complete and then copies data back to host memory before the constructor returns. Any command groups with requirements to the same memory object cannot execute until the host accessor is destroyed as shown on [Table 6](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#fig:host-acc).

so based on this it may be okay to have to simultaneous `host_accessor` to the same memory without it dead-locking. @gmlueck | @keryell - Is that your reading of the spec too or should it be UB?

Reply:
You are asking whether code like this could deadlock?

```
buffer b{/*...*/};
/* submit commands using accessor to "b" */
host_accessor ha1{b};
host_accessor ha2{b};  // does this deadlock because "ha1" is already accessing this buffer?
```

Interestingly, we actually just changed an example which inadvertently had code like this because we thought the example would deadlock.  (https://github.com/KhronosGroup/SYCL-Docs/pull/374)  However, I think your reading of the spec is correct.  It says that the host accessor constructor blocks waiting for *kernels* to complete.  It does not say it blocks waiting for other host accessors to be destroyed.

I don't know if we specifically thought of this case when we wrote the spec, but I think the spec does say that the code snippet I show above should not deadlock.

That said, this test is not trying to test the corner cases of host accessors.  Do we want to rely on this edge case here?  It might be better to add a test like this specifically to a test for host accessors, where the intent of the test is more clear.

Reply:
Therefore, I have to keep tests using `host_accessor` type

Reply:
For me the code above dead-lock because at least one of the accessor is writing. The same example with only read accessors would not dead-lock, according to usual data-dependency rules of SYCL.

Reply:
> That said, this test is not trying to test the corner cases of host accessors. Do we want to rely on this edge case here? It might be better to add a test like this specifically to a test for host accessors, where the intent of the test is more clear.

I agree with @gmlueck on this. It would be better to avoid this for unrelated testing and we can have a separate case for this where appropriate.